core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1158,74 +1158,91 @@ public int getNumberOfPinnedFiles() {
       AccessControlException {
+    List<Long> blocks;
     Metrics.DELETE_PATHS_OPS.inc();
     try (JournalContext journalContext = createJournalContext();
          LockedInodePath inodePath = mInodeTree.lockFullInodePath(path, InodeTree.LockMode.WRITE)) {
       mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
       mMountTable.checkUnderWritableMountPoint(path);
-      deleteAndJournal(inodePath, options, journalContext);
+      blocks = deleteAndJournal(inodePath, options, journalContext);
     }
+    mBlockMaster.removeBlocks(blocks, true /* delete */);
   }
 
   /**
    * Deletes a given path.
    * <p>
    * Writes to the journal.
    *
+   * This method does not delete blocks. Instead, it returns the blocks so that they can be deleted
+   * after the path deletion journal entry has been written. We cannot delete blocks earlier because
+   * the path deletion may fail, leaving us with files containing deleted blocks.
+   *
    * @param inodePath the path to delete
    * @param deleteOptions the method options
    * @param journalContext the journal context
+   * @return the list of deleted block ids
    * @throws InvalidPathException if the path is invalid
    * @throws FileDoesNotExistException if the file does not exist
    * @throws DirectoryNotEmptyException if recursive is false and the file is a nonempty directory
    */
-  private void deleteAndJournal(LockedInodePath inodePath, DeleteOptions deleteOptions,
+  private List<Long> deleteAndJournal(LockedInodePath inodePath, DeleteOptions deleteOptions,
       JournalContext journalContext) throws InvalidPathException, FileDoesNotExistException,
       IOException, DirectoryNotEmptyException {
     long opTimeMs = System.currentTimeMillis();
-    deleteInternal(inodePath, false, opTimeMs, deleteOptions, journalContext);
+    return deleteInternal(inodePath, false, opTimeMs, deleteOptions, journalContext);
   }
 
   /**
    * @param entry the entry to use
    */
   private void deleteFromEntry(DeleteFileEntry entry) {
     Metrics.DELETE_PATHS_OPS.inc();
-    try (LockedInodePath inodePath = mInodeTree
-        .lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
-      deleteInternal(inodePath, true, entry.getOpTimeMs(),
-          DeleteOptions.defaults().setRecursive(entry.getRecursive())
-              .setAlluxioOnly(entry.getAlluxioOnly()), NoopJournalContext.INSTANCE);
+    List<Long> blocks;
+    try (LockedInodePath inodePath =
+        mInodeTree.lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
+      blocks = deleteInternal(
+          inodePath, true, entry.getOpTimeMs(), DeleteOptions.defaults()
+              .setRecursive(entry.getRecursive()).setAlluxioOnly(entry.getAlluxioOnly()),
+          NoopJournalContext.INSTANCE);
+
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
+    mBlockMaster.removeBlocks(blocks, true /* delete */);
```

Comment:
Is this only to pick up the case where we journaled a delete file but not the corresponding delete block?